inference-engine/src/mkldnn_plugin/nodes/mkldnn_depth_to_space_node.cpp

```diff
@@ -0,0 +1,353 @@
+        THROW_ERROR << "has incorrect block_size parameter is zero!";
+
+    size_t nSpatialDims = srcDims.size() - 2;
+    blockStep = static_cast<size_t>(std::pow(blockSize, nSpatialDims));
+    if (srcDims[1] % blockStep)
+        THROW_ERROR << "has block_size parameter which is incompatible with input tensor channels dimension size";
+
+    if (srcDims[1] / blockStep != dstDims[1])
+        THROW_ERROR << "has incompatible input/output channels";
+
+    for (size_t i = 0; i < nSpatialDims; ++i) {
+        if (srcDims[i + 2] * blockSize != dstDims[i + 2])
+            THROW_ERROR << "has incompatible spatial dims";
+    }
+
+    if (getParentEdges().size() != 1)
+        THROW_ERROR << "has incorrect number of input edges";
+    if (getChildEdges().empty())
+        THROW_ERROR << "has incorrect number of output edges";
+}
+
+void MKLDNNDepthToSpaceNode::initSupportedPrimitiveDescriptors() {
+    if (!supportedPrimitiveDescriptors.empty())
+        return;
+
+    InferenceEngine::Precision precision = getCnnLayer()->insData[0].lock()->getPrecision();
+    auto dataType = MKLDNNExtensionUtils::IEPrecisionToDataType(precision);
+
+    auto srcDims = getParentEdgeAt(0)->getDims();
+    int nDims = srcDims.ToSizeVector().size();
+
+    InferenceEngine::LayerConfig config;
+    config.dynBatchSupport = true;
+    config.inConfs.resize(1);
+    config.outConfs.resize(1);
+    config.inConfs[0].inPlace = -1;
+    config.inConfs[0].constant = false;
+    config.outConfs[0].inPlace = -1;
+    config.outConfs[0].constant = false;
+
+    auto pushSupportedPrimitiveDescriptor = [&](const memory::format_tag memoryFormat) {
+        config.inConfs[0].desc = MKLDNNMemoryDesc(getParentEdgeAt(0)->getDims(), dataType, memoryFormat);
+        config.outConfs[0].desc = MKLDNNMemoryDesc(getChildEdgeAt(0)->getDims(), dataType, memoryFormat);
+        supportedPrimitiveDescriptors.push_back({config, impl_desc_type::unknown, memoryFormat});
+    };
+
+    auto canUseBlocked = [=](const size_t block) {
+        return srcDims[1] % block == 0 && (mode == Mode::BLOCKS_FIRST ? (srcDims[1] / block) % blockStep == 0 : block % blockStep == 0);
+    };
+
+    if (nDims == 4) {
+        pushSupportedPrimitiveDescriptor(mkldnn::memory::format_tag::nhwc);
+        if (canUseBlocked(8))
+            pushSupportedPrimitiveDescriptor(mkldnn::memory::format_tag::nChw8c);
+        if (canUseBlocked(16))
+            pushSupportedPrimitiveDescriptor(mkldnn::memory::format_tag::nChw16c);
+    } else if (nDims == 5) {
+        pushSupportedPrimitiveDescriptor(mkldnn::memory::format_tag::ndhwc);
+        if (canUseBlocked(8))
+            pushSupportedPrimitiveDescriptor(mkldnn::memory::format_tag::nCdhw8c);
+        if (canUseBlocked(16))
+            pushSupportedPrimitiveDescriptor(mkldnn::memory::format_tag::nCdhw16c);
+    }
+    pushSupportedPrimitiveDescriptor(MKLDNNMemory::GetPlainFormat(getParentEdgeAt(0)->getDims()));
+}
+
+void MKLDNNDepthToSpaceNode::createPrimitive() {
```

Comment:
@a-sidorova  Reading createPrimitive code I see 3 calls: `prepareParams(...)`, `prepareOptimizedParams(...)`, `prepareConfigParams(...)` and I give up to understand what they are doing.
To be honest I don't like the idea of inheritance from `PermuteUtils` at all.
What I would prefer to see is:
1) `MKLDNNDepthToSpaceNode::createPrimitive()` calls `PermuteUtils::PermuteUtils(static_params)` ctor where `static_params` is united structure that can be reused both by DepthToSpace and Permute itself.
2) `MKLDNNDepthToSpaceNode::execute()` calls `PermuteUtils::execute(dynamic_params)` method, where `dynamic_params` contains all parameters not available on load stage (like data pointers).
3) Profit

Is this approach feasible? Lets discuss offline if you have any concerns.

@chenhu-wang  What so you think?

Reply:
1. I agree that composite is better than inherit here, one benefit is we can explicitly see prepareParams(...) is from DepthAndSpaceUtils, and prepareOptimizedParams(...), prepareConfigParams(...) are from PermuteUtils. And the function name could be modified to self-explanation also.  such as prepareOptimizedParams(...) -> createPermute()
2. Seems like some case in block layout is not executed with permute. I think it's because these case can not be transformed to permute, not because permute do not support some case, because channel dimension is rearranged and not continuous any more. We need compare permute_reference  with reorder+permure_kernel.

Reply:
@dmitry-gorokhov I rewrote this class. Check please